File: frontend/src/lib/journey.ts
import type { Flight, Journey, JourneyAirport, JourneyLeg, JourneyMilestone, ResolveResult, Waypoint } from "../types";
import { gcDistanceNm } from "./geo";
import { num, routeIdents } from "./foreflight";

type ResolveFn = (idents: string[]) => ResolveResult;
const TIERS = [10, 25, 50, 100];

const personName = (raw?: string | null) => (raw ? String(raw).split(";")[0].trim() : "");

export function buildJourney(flights: Flight[], resolve: ResolveFn): Journey {
  const airports = new Map<string, JourneyAirport>();
  const legs = new Map<string, JourneyLeg>();
  const aircraftCounts = new Map<string, number>();
  const people = new Set<string>();
  const countries = new Set<string>();
  let totalHours = 0;
  let nightHours = 0;
  let firstDate: string | null = null;
  let longestLeg = 0;

  for (const f of flights) {
    totalHours += num(f.total_time);
    nightHours += num(f.night);
    const ac = (f.aircraft_id ?? "").trim();
    if (ac) aircraftCounts.set(ac, (aircraftCounts.get(ac) ?? 0) + 1);
    for (let n = 1; n <= 6; n++) {
      const nm = personName(f[`person${n}`] as string | null | undefined);
      if (nm) people.add(nm);
    }
    const d = f.date ?? null;
    if (d && (firstDate === null || d < firstDate)) firstDate = d;
  }

  const ordered = [...flights].sort((a, b) => (a.date ?? "").localeCompare(b.date ?? ""));
  const seenAirports = new Set<string>();
  const seenCountries = new Set<string>();
  const tierDates = new Map<number, string | null>();
  let firstIntlDate: string | null = null;

  for (const f of ordered) {
    const idents = routeIdents(f);
    const resolved = resolve(idents);
    const pts = idents.map((i) => resolved[i]).filter((p): p is Waypoint => !!p);
    if (pts.length < 2) continue;
    const d = f.date ?? null;
    const src = f.from_ ? resolved[f.from_] : null;
    const dst = f.to ? resolved[f.to] : null;

    // Airports = resolved airport-type endpoints, counted at most once per
    // flight (a round-trip like KSBA→KSBA is a single visit, per the spec).
    const endpoints = new Map<string, Waypoint>();
    for (const p of [src, dst]) if (p && p.type === "airport") endpoints.set(p.ident, p);
    for (const p of endpoints.values()) {
      let rec = airports.get(p.ident);
      if (!rec) {
        rec = { ident: p.ident, name: p.name ?? null, country: p.country ?? null, lat: p.lat, lon: p.lon, visits: 0, first_date: null, last_date: null };
        airports.set(p.ident, rec);
      }
      rec.visits += 1;
      if (d) {
        if (rec.first_date === null || d < rec.first_date) rec.first_date = d;
        if (rec.last_date === null || d > rec.last_date) rec.last_date = d;
      }
      if (p.country) countries.add(p.country);
    }

    if (src && dst) {
      const key = `${src.ident} ${dst.ident}`;
      let leg = legs.get(key);
      if (!leg) {
        leg = { from_ident: src.ident, to_ident: dst.ident, from_lat: src.lat, from_lon: src.lon, to_lat: dst.lat, to_lon: dst.lon, count: 0, last_date: null };
        legs.set(key, leg);
      }
      leg.count += 1;
      if (d && (leg.last_date === null || d > leg.last_date)) leg.last_date = d;
      longestLeg = Math.max(longestLeg, gcDistanceNm(src.lat, src.lon, dst.lat, dst.lon));
    }

    for (const p of [src, dst]) if (p?.country) seenCountries.add(p.country);
    if (firstIntlDate === null && seenCountries.size >= 2) firstIntlDate = d;
    for (const p of [src, dst]) if (p?.type === "airport") seenAirports.add(p.ident);
    for (const tier of TIERS) if (!tierDates.has(tier) && seenAirports.size >= tier) tierDates.set(tier, d);
  }

  const airportList = [...airports.values()];
  const homeBase = airportList.reduce<JourneyAirport | null>((best, a) => (!best || a.visits > best.visits ? a : best), null);
  let furthest = 0;
  if (homeBase) {
    for (const a of airportList) {
      if (a.ident === homeBase.ident) continue;
      furthest = Math.max(furthest, gcDistanceNm(homeBase.lat, homeBase.lon, a.lat, a.lon));
    }
  }
  let callsign: string | null = null;
  let bestCount = -1;
  for (const [ac, c] of aircraftCounts) if (c > bestCount) { bestCount = c; callsign = ac; }

  const legList = [...legs.values()];
  const longestLegObj = legList.reduce<JourneyLeg | null>(
    (best, lg) => {
      const dist = gcDistanceNm(lg.from_lat, lg.from_lon, lg.to_lat, lg.to_lon);
      const bestDist = best ? gcDistanceNm(best.from_lat, best.from_lon, best.to_lat, best.to_lon) : -1;
      return dist > bestDist ? lg : best;
    },
    null,
  );

  const milestones: JourneyMilestone[] = [];
  if (flights.length) milestones.push({ key: "first_flight", label: "First flight", date: firstDate });
  if (countries.size >= 2) milestones.push({ key: "first_international", label: "First international", date: firstIntlDate });
  for (const tier of TIERS) if (airports.size >= tier) milestones.push({ key: `airports_${tier}`, label: `${tier} airports`, date: tierDates.get(tier) ?? null });
  if (longestLegObj) milestones.push({ key: "longest_journey", label: "Longest journey", date: longestLegObj.last_date });

  const r1 = (n: number) => Math.round(n * 10) / 10;
  return {
    airports: airportList,
    legs: legList,
    identity: {
      callsign,
      first_flight_year: firstDate ? firstDate.slice(0, 4) : null,
      home_base: homeBase ? homeBase.ident : null,
      total_hours: r1(totalHours),
      airport_count: airports.size,
      country_count: countries.size,
      aircraft_count: aircraftCounts.size,
      people_count: people.size,
      night_hours: r1(nightHours),
      longest_leg_nm: Math.round(longestLeg),
      furthest_from_home_nm: Math.round(furthest),
    },
    milestones,
  };
}

File: backend/app/coords/journey.py
"""Aggregate a list of flights into the Journey payload.

Resolution happens per flight (the resolver disambiguates colliding idents by
the airports of a single flight), then results are unioned into airport, leg,
identity, and milestone collections. The TS mirror lives in
frontend/src/lib/journey.ts — keep the two in sync (see the shared contract in
docs/superpowers/plans/2026-05-31-journey-page.md)."""
from collections.abc import Callable, Iterable

from app.coords.geo import gc_distance_nm

ResolveFn = Callable[[list[str]], dict[str, dict | None]]

_TIERS = [10, 25, 50, 100]


def _f(v) -> float:
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0.0


def _idents(f: dict) -> list[str]:
    mid = (f.get("route") or "").split()
    return [x for x in [f.get("from_"), *mid, f.get("to")] if x]


def _person_name(raw) -> str:
    if not raw:
        return ""
    return str(raw).split(";")[0].strip()


def build_journey(flights: Iterable[dict], resolve: ResolveFn) -> dict:
    flights = list(flights)

    airports: dict[str, dict] = {}        # ident -> airport record
    legs: dict[tuple[str, str], dict] = {}
    aircraft_counts: dict[str, int] = {}
    people: set[str] = set()
    countries: set[str] = set()
    total_hours = night_hours = 0.0
    first_date: str | None = None
    longest_leg = 0.0

    # For milestone date tracking we iterate oldest first.
    def datekey(f):
        return f.get("date") or ""

    ordered = sorted(flights, key=datekey)

    seen_airports: set[str] = set()
    seen_countries: set[str] = set()
    airport_tier_dates: dict[int, str | None] = {}
    first_intl_date: str | None = None

    for f in flights:
        total_hours += _f(f.get("total_time"))
        night_hours += _f(f.get("night"))
        ac = (f.get("aircraft_id") or "").strip()
        if ac:
            aircraft_counts[ac] = aircraft_counts.get(ac, 0) + 1
        for n in range(1, 7):
            nm = _person_name(f.get(f"person{n}"))
            if nm:
                people.add(nm)
        d = f.get("date")
        if d and (first_date is None or d < first_date):
            first_date = d

    for f in ordered:
        idents = _idents(f)
        resolved = resolve(idents)
        pts = [resolved.get(i) for i in idents]
        pts = [p for p in pts if p]
        if len(pts) < 2:
            continue
        d = f.get("date")

        # Endpoints (first/last *resolved* of from_/to specifically).
        src = resolved.get(f.get("from_") or "")
        dst = resolved.get(f.get("to") or "")

        # Airports = resolved airport-type endpoints, counted at most once per
        # flight (a round-trip like KSBA→KSBA is a single visit, per the spec).
        endpoints = {}
        for p in (src, dst):
            if p and p.get("type") == "airport":
                endpoints[p["ident"]] = p
        for p in endpoints.values():
            rec = airports.get(p["ident"])
            if not rec:
                rec = {
                    "ident": p["ident"], "name": p.get("name"), "country": p.get("country"),
                    "lat": p["lat"], "lon": p["lon"], "visits": 0,
                    "first_date": None, "last_date": None,
                }
                airports[p["ident"]] = rec
            rec["visits"] += 1
            if d:
                if rec["first_date"] is None or d < rec["first_date"]:
                    rec["first_date"] = d
                if rec["last_date"] is None or d > rec["last_date"]:
                    rec["last_date"] = d
            if p.get("country"):
                countries.add(p["country"])

        # Leg keyed by resolved from/to endpoints.
        if src and dst:
            key = (src["ident"], dst["ident"])
            leg = legs.get(key)
            if not leg:
                leg = {
                    "from_ident": src["ident"], "to_ident": dst["ident"],
                    "from_lat": src["lat"], "from_lon": src["lon"],
                    "to_lat": dst["lat"], "to_lon": dst["lon"],
                    "count": 0, "last_date": None,
                }
                legs[key] = leg
            leg["count"] += 1
            if d and (leg["last_date"] is None or d > leg["last_date"]):
                leg["last_date"] = d
            longest_leg = max(longest_leg, gc_distance_nm(src["lat"], src["lon"], dst["lat"], dst["lon"]))

        # Milestone bookkeeping (oldest-first).
        flight_countries = {p.get("country") for p in (src, dst) if p and p.get("country")}
        for c in flight_countries:
            seen_countries.add(c)
        if first_intl_date is None and len(seen_countries) >= 2:
            first_intl_date = d
        for p in (src, dst):
            if p and p.get("type") == "airport":
                seen_airports.add(p["ident"])
        for tier in _TIERS:
            if tier not in airport_tier_dates and len(seen_airports) >= tier:
                airport_tier_dates[tier] = d

    airport_list = list(airports.values())
    home_base = max(airport_list, key=lambda a: a["visits"], default=None)
    home_base_ident = home_base["ident"] if home_base else None

    furthest = 0.0
    if home_base:
        for a in airport_list:
            if a["ident"] == home_base_ident:
                continue
            furthest = max(furthest, gc_distance_nm(home_base["lat"], home_base["lon"], a["lat"], a["lon"]))

    callsign = max(aircraft_counts, key=aircraft_counts.get, default=None) if aircraft_counts else None

    leg_list = list(legs.values())
    longest_leg_obj = max(leg_list, key=lambda lg: gc_distance_nm(lg["from_lat"], lg["from_lon"], lg["to_lat"], lg["to_lon"]), default=None)

    milestones: list[dict] = []
    if first_date is not None or flights:
        milestones.append({"key": "first_flight", "label": "First flight", "date": first_date})
    if len(countries) >= 2:
        milestones.append({"key": "first_international", "label": "First international", "date": first_intl_date})
    for tier in _TIERS:
        if len(airports) >= tier:
            milestones.append({"key": f"airports_{tier}", "label": f"{tier} airports", "date": airport_tier_dates.get(tier)})
    if longest_leg_obj is not None:
        milestones.append({"key": "longest_journey", "label": "Longest journey", "date": longest_leg_obj["last_date"]})

    return {
        "airports": airport_list,
        "legs": leg_list,
        "identity": {
            "callsign": callsign,
            "first_flight_year": first_date[:4] if first_date else None,
            "home_base": home_base_ident,
            "total_hours": round(total_hours, 1),
            "airport_count": len(airports),
            "country_count": len(countries),
            "aircraft_count": len(aircraft_counts),
            "people_count": len(people),
            "night_hours": round(night_hours, 1),
            "longest_leg_nm": round(longest_leg),
            "furthest_from_home_nm": round(furthest),
        },
        "milestones": milestones,
    }
